### PR TITLE
[com_fields] Adding label Class for input

### PR DESCRIPTION
**Pull request by coolcat-creations**
Pull Request for Issue #16994.
## Summary of Changes
Removed the hide Label Option Replaced hide Label Option with a label Class Option. Added B/C Option to add class hidden to the class if label was set previously to hidden. Added a label class Option for Form Reordered the Options and added headlines.
See also discussion in #17177
## Testing Instructions
	1.	Hide your label, apply the patch and see if the label is still hidden
	2.	Add a class for the label in the form and see if it's added
	3.	Add a class for the label in the output rendering and see if it's added
## Documentation Changes Required
Yes adding the possibility to add a Label Class in the form.

grafik

grafik

### CODE REVIEW

**Comment by ciar4n** [comment]
A slight B/C issue. If the 'Show Label' has been set to hide, once the PR is applied there is no way to un-hide the label. The hidden class is saved to the user data with no way to remove it.

**Comment by coolcat-creations** [comment]
@ciar4n good point. Any other idea how to make it B/C able?
hmmm...

**Comment by ciar4n** [comment]
In this case none that I can think of.

I think for now we could keep the 'Show Label' fields. Considering the 'Render Class' exists which is adding a class to the field container, I would question the need for a 'Label Class'. So my 2 cents would be to revert replacing the 'Show Label' fields with the 'Label Class' fields.

**Comment by coolcat-creations** [comment]
Yes, you are right, we can control the label by the render class. Just tried to be more consistent for both places. Will change it.

**Comment by coolcat-creations** [comment]
Ok, looks like that now:
grafik

**Comment by coolcat-creations** [comment]
Looks like that now:
grafik

**Comment by infograf768** [comment]
This works fine, but I really would like to get a Label Class when rendering.
What we have now is limited to all fields `field-label`

**Comment by coolcat-creations** [comment]
@infograf768 not sure what you mean? Can you try to explain it different, what the issue is?

**Comment by infograf768** [comment]
In this patch, the Label Class is used when editing an item in backend.
Example adding class invalid to label:
screen shot 2017-11-02 at 17 15 24

gives when editing an article in backend:
screen shot 2017-11-02 at 17 16 24

But this can't have any effect on frontend
screen shot 2017-11-02 at 17 17 52

To be able to render a specific class in frontend, you need a new field in the xml, and then modify the files
`ROOT/components/com_contact/layouts/field/render.php`
and
`/ROOT/components/com_fields/layouts/field/render.php`

to add that class after `field-label`

**Comment by infograf768** [comment]
That new field could be displayed with a showon depending on the value of the Show Label field

**Comment by coolcat-creations** [comment]
@infograf768 i still don't understand. The label class is only for the form (input) not for the rendering part. If you look at the options I even added headlines to clarify this. 

In the first version of the PR i did quite exactly what you are suggesting but we said then to remove the label class for frontend again and to leave the show / hide field for rendering.

**Comment by infograf768** [comment]
I still think it is quite helpful to be able to add a specific class in frontend for the label rendering when rendering Label is set to Yes. This will let using different rendering depending on the field.

The B/C issue pointed at by @ciar4n does not exist as far as I could test.
A field formerly set to Hide Label let's change it to show after my patch.

When hiding label, the render label class is just not used as the label is hidden.

This is what I have done:
Added a new field in in `/administrator/components/com_fields/models/forms/field.xml` (the lang strings are already in your PR) after `Show Label`:

```
			<field
				name="render_label_class"
				type="textarea"
				label="COM_FIELDS_FIELD_LABEL_RENDER_CLASS_LABEL"
				description="COM_FIELDS_FIELD_LABEL_RENDER_CLASS_DESC"
				class="input-xxlarge"
				size="40"
				showon="showlabel:1"
			/>
```
Added the node in fieldsplugin.php line 174 (not sure that one is necessary)
`$node->setAttribute('renderlabelclass', $field->params->get('render_label_class'));`

Modified `/components/com_fields/layouts/field/render.php` to be able to use the `render_label_class`

```
if (!key_exists('field', $displayData))
{
	return;
}

$field = $displayData['field'];
$label = JText::_($field->label);
$value = $field->value;
$showLabel = $field->params->get('showlabel');
$renderLabelClass = $field->params->get('render_label_class');

if ($value == '')
{
	return;
}

?>
<?php if ($showLabel == 1) : ?>
	<span class="field-label <?php echo $renderLabelClass; ?>"><?php echo htmlentities($label, ENT_QUOTES | ENT_IGNORE, 'UTF-8'); ?>: </span>
<?php endif; ?>
<span class="field-value"><?php echo $value; ?></span>
```
Same has to be done for the specific `/components/com_contact/layouts/field/render.php`

I get:
customfieldslabelclass


EDIT: if I un-hide the Label, it works fine.

**Comment by ciar4n** [comment]
The B/C issue has been fixed with https://github.com/joomla/joomla-cms/pull/18431/commits/741c13da45c2316b023390a4f238c4adca642aa4

**Comment by infograf768** [comment]
With my patch, as explained above, I could not reproduce your B/C issue, which means we can add the render_label_class as suggested.

**Comment by infograf768** [comment]
If you like, I can show you with animated screenshot

**Comment by infograf768** [comment]
@ciar4n 
Shall I do these captures?

**Comment by coolcat-creations** [comment]
@infograf768 you can define the class of the label by .renderclass label thats why we dont add another additional parameter. We start adding too much parameters again. This PR is just to be able to add a class for the label in the form in order to hide it or to do different things. it would be better of course to have the same for the Rendering too but there is already a show/hide field. Means we would need to add another field. As far i know we wanted to prevent to add too many options?

**Comment by ciar4n** [comment]
My apologies.. I thought your b/c comment was on the PR in general rather than your specific suggestion. As your suggestion doesn't remove the 'Show Label' field, I see no b/c issue with it.

Adding a 'Render Label Class' certainly makes things a little easier for alternative styling on the label. My only thought on it is that in the case of the render field, you could use the 'Render Class' field to do the same thing. The difference been that you will need to create custom CSS for the class rather than been able to use a currently existing bootstrap class.

``` CSS
.my-label-class label {
   // New label styling
}
```

**Comment by coolcat-creations** [comment]
or 
.my-containerclass label + label {
}

 it's just not a label in the rendering, i think it's a span

**Comment by infograf768** [comment]
In fact, it would rather be
```
.field-label .my-label-class {
    // style
}
```
to target specifically the custom field markup as we would have
`<span class="field-label <?php echo $renderLabelClass; ?>"`

In any case, I really think we need this even if it is a new parameter.
@coolcat-creations 
Would you accept to update your PR?

**Comment by coolcat-creations** [comment]
I can add back the class of course. that was initially what i wanted to do. Just did not want to overcrowd the parameters again.

**Comment by infograf768** [comment]
Thanks. Will test when done.

**Comment by coolcat-creations** [comment]
Done :)

File: administrator/components/com_fields/models/forms/field.xml [hunk 1]
```diff
@@ -298,6 +298,16 @@
 			</field>
 
 			<field
+					name="label_render_class"
```
Comment (infograf768) [inline administrator/components/com_fields/models/forms/field.xml]: 1 tab too much for the new field `name` etc..

**Comment by infograf768** [comment]
Also missing same changes in `/components/com_contact/layouts/field/render.php`

**Comment by coolcat-creations** [comment]
oops,  was too fast doing the stuff and did not commit all files. sorrrryy

**Comment by infograf768** [comment]
take your time, I am leaving for weekend

File: administrator/components/com_fields/models/forms/field.xml [hunk 2]
```diff
@@ -276,6 +271,43 @@
 			</field>
 
 			<field
+				name="renderoptions"
+				type="spacer"
```
Comment (Quy) [inline administrator/components/com_fields/models/forms/field.xml]: Use `note` type so you don't need `<strong>` markup in the language string.

File: administrator/components/com_fields/models/forms/field.xml [hunk 3]
```diff
@@ -223,6 +223,13 @@
 
 	<fields name="params" label="COM_FIELDS_FIELD_BASIC_LABEL">
 		<fieldset name="basic">
+
+			<field
+				name="formoptions"
+				type="spacer"
```
Comment (Quy) [inline administrator/components/com_fields/models/forms/field.xml]: Use `note` type so you don't need `<strong>` markup in the language string.

File: administrator/components/com_fields/models/forms/field.xml [hunk 4]
```diff
@@ -276,6 +271,43 @@
...
+			/>
+
+			<field
+				name="showlabel"
+				type="radio"
+				label="COM_FIELDS_FIELD_SHOWLABEL_LABEL"
+				description="COM_FIELDS_FIELD_SHOWLABEL_DESC"
+				class="btn-group btn-group-yesno"
+				default="1"
+			>
```
Comment (Quy) [inline administrator/components/com_fields/models/forms/field.xml]: Indent.

**Comment by infograf768** [comment]
Agree with @Quy concerning note which means modifying both strings.

**Comment by infograf768** [comment]
@coolcat-creations 
can you do the last desired modifications so that we can test positively?

**Comment by coolcat-creations** [comment]
Thank you, just did the changes

**Comment by infograf768** [comment]
@coolcat-creations 
still missing the indent remarked by @Quy line 295. See above.

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 7351e21674b50486b6205db0cf1add1fd2011312<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18431">issues.joomla.org/tracker/joomla-cms/18431</a>.</sub>

File: administrator/language/en-GB/en-GB.com_fields.ini [hunk 5]
```diff
@@ -31,6 +33,10 @@ COM_FIELDS_FIELD_IMAGE_LABEL="Image"
 COM_FIELDS_FIELD_INVALID_DEFAULT_VALUE="The default value is invalid."
 COM_FIELDS_FIELD_LABEL_DESC="The label of the field to display."
 COM_FIELDS_FIELD_LABEL_LABEL="Label"
+COM_FIELDS_FIELD_LABEL_FORM_CLASS_LABEL="Label Class"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_fields.ini]: Alphabetize keys from lines 35-39.

Comment (infograf768) [reply administrator/language/en-GB/en-GB.com_fields.ini]: oops, did not check that one. Good find

Comment (coolcat-creations) [reply administrator/language/en-GB/en-GB.com_fields.ini]: Thank you, learned something new :)

**Comment by JeroenMoolenschot** [comment]
I have tested this item :white_check_mark: successfully on 9721f9f402927f4e5de0faf0b3e2b81ac5be88ab<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18431">issues.joomla.org/tracker/joomla-cms/18431</a>.</sub>

**Comment by ghost** [comment]
Ready to Commit after two successful tests.

**Review by brianteeman — APPROVED** [review]

**Pull request merged**